### DISCUSSION TITLE
Inherited table with shared columns but different field names

### DISCUSSION THREAD

**Discussion by fanzeyi** [sqlalchemy/sqlalchemy#9330] [Usage Questions] [answered]
So I'm running into a bit issue when trying to create some polymorphic classes with SQLAlchemy.

Specifically, I'm trying to have two sub-classes sharing the same column that these sub-classes declared themselves, while having a different field name.

For example (taken from SQLAlchemy document):

```python
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase):
    pass


class Employee(Base):
    __tablename__ = "employee"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    type: Mapped[str]

    __mapper_args__ = {
        "polymorphic_identity": "employee",
        "polymorphic_on": "type",
    }

    def __repr__(self):
        return f"{self.__class__.__name__}({self.name!r})"


class Engineer(Employee):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }

    team: Mapped[str] = mapped_column(
        "team", nullable=True, use_existing_column=True
    )


class Manager(Employee):
    __mapper_args__ = {
        "polymorphic_identity": "manager",
    }

    department: Mapped[str] = mapped_column(
        "team", nullable=True, use_existing_column=True
    )
```

I want `Manager.department` to share the same column as `Engineer.team` in the table. Even though I have specified them to be the same column name, SQLAlchemy doesn't seem to recognize that and it gives me:

```
ArgumentError: Column 'team' on class Manager conflicts with existing column 'employee.team'.  If using Declarative, consider using the use_existing_column parameter of mapped_column() to resolve conflicts.
```

However `use_existing_column=True` is already set.

I traced the code a bit and it looks like here, SQLAlchemy doesn't really consider the case where column name is explicitly specified:

https://github.com/sqlalchemy/sqlalchemy/blob/c6ecc39b62b41c4f21375a51bac582c61baa3d9d/lib/sqlalchemy/orm/properties.py#L666-L668

What do you think? Should SQLAlchemy be able to support this use case?

**Comment by zzzeek** [comment, marked as answer]
hi yes , this is a bug , I never considered this use.

you can use the old method for now, which you would just need to have `# type: ignore` to keep type checkers happy, which you can see in the 1.4 docs at :


https://docs.sqlalchemy.org/en/14/orm/inheritance.html#resolving-column-conflicts

I just wrote it out below to make sure it works

```py
from sqlalchemy import ForeignKey, Column, select, String
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    mapped_column,
    declared_attr,
)


class Base(DeclarativeBase):
    pass


class Employee(Base):
    __tablename__ = "employee"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    type: Mapped[str]

    __mapper_args__ = {
        "polymorphic_identity": "employee",
        "polymorphic_on": "type",
    }

    def __repr__(self):
        return f"{self.__class__.__name__}({self.name!r})"


class Engineer(Employee):
    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }

    @declared_attr
    def team(cls) -> Mapped[str]:
        return Employee.__table__.c.get(  # type: ignore
            "team", mapped_column(String, nullable=True)  # type: ignore
        )


class Manager(Employee):
    __mapper_args__ = {
        "polymorphic_identity": "manager",
    }

    @declared_attr
    def department(cls) -> Mapped[str]:
        return Employee.__table__.c.get( # type: ignore
            "team", mapped_column(String, nullable=True)  # type: ignore
        )


print(select(Engineer.team))
print(select(Manager.department))

```

opened #9332

**Reply by fanzeyi** [reply]
Thank you for the workaround and fix!

**Comment by zaksmolen** [comment]
Hi! I just wanted to check if there is any way to do this in the latest version or if this is still the best way? I have a similar situation where each child class has a single bool field that indicates something different, but I'd like to share the space in the DB.

**Reply by zzzeek** [reply]
the issue is mentioned above, its #9332 and it's fixed, was released in early 2.0 versions.  the example given at the top of this document works as written.
